cmake/Modules/ValkeySetup.cmake

```
@@ -89,6 +89,11 @@ macro (valkey_build_and_install_bin target sources ld_flags libs link_name)
         target_link_libraries(${target} OpenSSL::SSL valkey::valkey_tls)
     endif ()
 
+    if (USE_RDMA)
+        # Add required libraries needed for RDMA
+        target_link_libraries(${target} valkey::valkey_rdma valkey::valkey)
```

`valkey::valkey` is already added a few lines above. No need to add it again, right?

```suggestion
        target_link_libraries(${target} valkey::valkey_rdma)
```

Btw, I don't understand why all the binaries use the same libraries? valkey-cli, valkey-benchmark and valkey-server don't necessarily need the same libraries. @eifrah-aws why is it like this? Please help.

---

Build this on Ubuntu 24.04 LTS, cmake version 3.28.3, gcc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0:
If removing `valkey::valkey`, hitting a compiling error as following message. We can see `./lib/libvalkey.a ../lib/libvalkey_rdma.a` are linked, but `vk_usec_now` can't be linked. @eifrah-aws do you have any suggestion?

```
make valkey-cli VERBOSE=1

/usr/bin/cc   -DUSE_JEMALLOC -DUSE_RDMA=1 -DBUILD_RDMA_MODULE=0 -std=gnu11 -pedantic "CMakeFiles/valkey-cli.dir/anet.c.o" "CMakeFiles/valkey-cli.dir/adlist.c.o" "CMakeFiles/valkey-cli.dir/dict.c.o" "CMakeFiles/valkey-cli.dir/sds.c.o" "CMakeFiles/valkey-cli.dir/sha256.c.o" "CMakeFiles/valkey-cli.dir/util.c.o" "CMakeFiles/valkey-cli.dir/valkey-cli.c.o" "CMakeFiles/valkey-cli.dir/zmalloc.c.o" "CMakeFiles/valkey-cli.dir/release.c.o" "CMakeFiles/valkey-cli.dir/ae.c.o" "CMakeFiles/valkey-cli.dir/serverassert.c.o" "CMakeFiles/valkey-cli.dir/crcspeed.c.o" "CMakeFiles/valkey-cli.dir/crccombine.c.o" "CMakeFiles/valkey-cli.dir/crc64.c.o" "CMakeFiles/valkey-cli.dir/siphash.c.o" "CMakeFiles/valkey-cli.dir/crc16.c.o" "CMakeFiles/valkey-cli.dir/monotonic.c.o" "CMakeFiles/valkey-cli.dir/cli_common.c.o" "CMakeFiles/valkey-cli.dir/mt19937-64.c.o" "CMakeFiles/valkey-cli.dir/strl.c.o" "CMakeFiles/valkey-cli.dir/cli_commands.c.o" -o ../bin/valkey-cli  ../jemalloc-build/lib/libjemalloc.a ../lib/libfpconv.a ../lib/liblinenoise.a -rdynamic -pthread -ldl -lm ../lib/libvalkey.a ../lib/libvalkey_rdma.a /usr/lib/x86_64-linux-gnu/librdmacm.so /usr/lib/x86_64-linux-gnu/libibverbs.so 
/usr/bin/ld: ../lib/libvalkey_rdma.a(rdma.c.o): in function `vk_msec_now':
rdma.c:(.text+0x377): undefined reference to `vk_usec_now'
collect2: error: ld returned 1 exit status

```

---

Since these dependencies are statically linked the order is important. If libvalkey_rdma.a uses code from libvalkey.a then libvalkey.a  need to be last in the link list.
CMake should be able to handle this ordering when calling the linker, but it seems that it doesn't know.

Looking at the CMakeLists.txt in libvalkey we could instead fix it by adding the dependency in deps/libvalkey/CMakeLists.txt:
`target_link_libraries(valkey_rdma LINK_PRIVATE valkey::valkey ${RDMACM_LIBRARIES} ${IBVERBS_LIBRARIES})`
We need to fix it upstream to, and probably change the usage of LINK_PRIVATE to make the transitive dependencies work when the libraries are installed (valkey includes the targets and uses the built libraries directly).

---

OK, good point. Let's add a code comment about it here.

```suggestion
        # Bug in libvalkey 0.1.0: The order is important and we need to link
        # valkey::valkey after valkey::valkey_rdma. This will be fixed upstream.
        # TODO: Next time we lift libvalkey, remove valkey::valkey from
        # the next line.
        target_link_libraries(${target} valkey::valkey_rdma valkey::valkey)
```

---

The issue at hand is addressed in https://github.com/valkey-io/libvalkey/pull/201